redbox-core/redbox/graph/nodes/processes.py

```diff
@@ -250,20 +270,45 @@ def build_tool_pattern(tools=list[StructuredTool]) -> Callable[[RedboxState], di
     def _tool(state: RedboxState) -> dict[str, Any]:
         state_updates: list[dict] = []
 
-        for tool_id, tool_call_dict in state.get("tool_calls", {}).items():
+        tool_calls = state.get("tool_calls", {})
+        if not tool_calls:
+            log.warning("No tool calls found in state")
+            return None
```

Comment:
```suggestion
            return {}
```
or change annotation?